Retail Sales Data Processing and Business Insights Generation

Company: ABC Retail Solutions

Objective:
Build an end-to-end data engineering pipeline to ingest,
clean, transform, validate, secure, and analyze retail
sales data for business intelligence reporting.

In [ ]:
from google.colab import files

uploaded = files.upload()

Saving USECASE - Data Engineering.xlsx to USECASE - Data Engineering (1).xlsx


In [ ]:
import pandas as pd

In [ ]:
file_path = "USECASE - Data Engineering.xlsx"

In [ ]:
product_df = pd.read_excel(
    file_path,
    sheet_name="PRODUCT DETAILS",
    header=3
)

retail1 = pd.read_excel(
    file_path,
    sheet_name="RETAIL DATA 1",
    header=2
)

retail2 = pd.read_excel(
    file_path,
    sheet_name="RETAIL DATA 2",
    header=4
)

In [ ]:
product_df.head()

,product_id,product_name,category,price
0,101,Laptop,Electronics,250000
1,102,Phone,Electronics,70000
2,103,Shirt,Clothing,1300
3,104,Shoes,Clothing,8000
4,105,TV,Electronics,45000


In [ ]:
retail1.head()

,transaction_id,customer_id,customer_name,product_id,price,product_name,category,purchase_location,city,transaction_date,quantity,payment_method,discount,email,phone,payment_status
0,1,642,Troy Mitchell,108,15000.0,Mixer grinder,Home Appliances,offline,Bangalore,2025-12-06 00:00:00,3,UPI,0.10,Troy60@gmail.com,8385276968,successful
1,2,881,Sarah Guerrero,102,70000.0,Phone,ELEC,online,Bangalore,2025-07-21 00:00:00,5,Card,0.35,SarahGuerrero@yahoo.com,7147248911,successful
2,3,505,Samantha Hull,109,65000.0,Refrigerator,Home Appliances,offline,Chennai,2025-07-11 00:00:00,2,UPI,0.05,SamanthaHull@outlook.com,7415321565,successful
3,4,794,Gerald Cooper,106,80000.0,sofa,FURN,offline,Chennai,2025-10-06 00:00:00,2,UPI,0.05,GeraldCooper@gmail.com,9739354201,successful
4,5,864,Cameron Black,108,15000.0,Mixer grinder,home appliances,offline,Chennai,2025-12-12 00:00:00,1,Cash,0.05,CameronBlack@yahoo.com,9938169477,successful


In [ ]:
retail2.head()

,transaction_id,customer_id,customer_name,product_id,price,product_name,category,purchase_location,city,transaction_date,quantity,payment_method,discount,email,phone,payment_status
0,4001,1112,Jasmine Carlson,106,80000.0,sofa,Furniture,offline,Delhi,12-18-2025,2,Cash,0.10,JasmineCarlson@gmail.com,8286662057,successful
1,4002,1125,Jeanne Suarez,105,45000.0,Tv,Electronics,online,Mumbai,2025-08-07 00:00:00,3,Card,0.10,JeanneSuarez@yahoo.com,6435924212,successful
2,4003,1379,Anthony Howard,107,50000.0,Dining table,FURN,offline,Delhi,09-19-2025,2,Cash,0.30,Anthony42@gmail.com,8187928333,successful
3,4004,1570,John Wilkins,107,50000.0,DINING TABLE,FURN,offline,Delhi,2025-11-20 00:00:00,2,Cash,0.40,JohnWilkins@outlook.com,6440039629,successful
4,4005,1433,Joseph Knapp,108,15000.0,Mixer grinder,home appliances,online,Bangalore,2025-09-29 00:00:00,5,Cash,0.25,JosephKnapp@gmail.com,9284221237,successful


In [ ]:
print("Product Details Shape:", product_df.shape)
print("Retail Data 1 Shape:", retail1.shape)
print("Retail Data 2 Shape:", retail2.shape)

Product Details Shape: (10, 4)
Retail Data 1 Shape: (4243, 16)
Retail Data 2 Shape: (4251, 16)


In [ ]:
print(product_df.columns)
print(retail1.columns)
print(retail2.columns)

Index(['product_id', 'product_name', 'category', 'price'], dtype='object')
Index(['transaction_id', 'customer_id', 'customer_name', 'product_id', 'price',
       'product_name', 'category', 'purchase_location', 'city',
       'transaction_date', 'quantity', 'payment_method', 'discount', 'email',
       'phone', 'payment_status'],
      dtype='object')
Index(['transaction_id', 'customer_id', 'customer_name', 'product_id', 'price',
       'product_name', 'category', 'purchase_location', 'city',
       'transaction_date', 'quantity', 'payment_method', 'discount', 'email',
       'phone', 'payment_status'],
      dtype='object')


In [ ]:
retail1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4243 entries, 0 to 4242
Data columns (total 16 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   transaction_id     4243 non-null   int64  
 1   customer_id        4243 non-null   int64  
 2   customer_name      4243 non-null   object 
 3   product_id         4243 non-null   int64  
 4   price              3839 non-null   float64
 5   product_name       4243 non-null   object 
 6   category           4243 non-null   object 
 7   purchase_location  4243 non-null   object 
 8   city               4243 non-null   object 
 9   transaction_date   4243 non-null   object 
 10  quantity           4243 non-null   int64  
 11  payment_method     4243 non-null   object 
 12  discount           4243 non-null   float64
 13  email              4243 non-null   object 
 14  phone              4243 non-null   int64  
 15  payment_status     4243 non-null   object 
dtypes: float64(2), int64(5),

In [ ]:
retail2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4251 entries, 0 to 4250
Data columns (total 16 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   transaction_id     4251 non-null   int64  
 1   customer_id        4251 non-null   int64  
 2   customer_name      4251 non-null   object 
 3   product_id         4251 non-null   int64  
 4   price              3846 non-null   float64
 5   product_name       4251 non-null   object 
 6   category           4251 non-null   object 
 7   purchase_location  4251 non-null   object 
 8   city               4251 non-null   object 
 9   transaction_date   4251 non-null   object 
 10  quantity           4251 non-null   int64  
 11  payment_method     4251 non-null   object 
 12  discount           4251 non-null   float64
 13  email              4251 non-null   object 
 14  phone              4251 non-null   int64  
 15  payment_status     4251 non-null   object 
dtypes: float64(2), int64(5),

In [ ]:
retail2.isnull().sum()

,0
transaction_id,0
customer_id,0
customer_name,0
product_id,0
price,405
product_name,0
category,0
purchase_location,0
city,0
transaction_date,0


In [ ]:
retail1.isnull().sum()

,0
transaction_id,0
customer_id,0
customer_name,0
product_id,0
price,404
product_name,0
category,0
purchase_location,0
city,0
transaction_date,0


In [ ]:
print(retail1.duplicated().sum())
print(retail2.duplicated().sum())

0
0


Data Quality Assessment

Findings:

1. Missing values identified in Price column.
2. transaction_date stored as object datatype.
3. Email and phone contain PII information.
4. Duplicate records need validation.
5. Product and category standardization required.

In [ ]:
print("Retail1 Missing Values")
print(retail1.isnull().sum())

print("\nRetail2 Missing Values")
print(retail2.isnull().sum())

print("\nRetail1 Duplicates")
print(retail1.duplicated().sum())

print("\nRetail2 Duplicates")
print(retail2.duplicated().sum())

Retail1 Missing Values
transaction_id         0
customer_id            0
customer_name          0
product_id             0
price                404
product_name           0
category               0
purchase_location      0
city                   0
transaction_date       0
quantity               0
payment_method         0
discount               0
email                  0
phone                  0
payment_status         0
dtype: int64

Retail2 Missing Values
transaction_id         0
customer_id            0
customer_name          0
product_id             0
price                405
product_name           0
category               0
purchase_location      0
city                   0
transaction_date       0
quantity               0
payment_method         0
discount               0
email                  0
phone                  0
payment_status         0
dtype: int64

Retail1 Duplicates
0

Retail2 Duplicates
0


Data Quality Assessment Results

1. Retail Data 1 contains 404 missing price values.
2. Retail Data 2 contains 405 missing price values.
3. No duplicate records were identified in either dataset.
4. transaction_date is stored as object datatype and requires conversion.
5. Email and phone columns contain PII information and require masking.
6. Product and category standardization will be performed.

In [ ]:
retail_df = pd.concat(
    [retail1, retail2],
    ignore_index=True
)

In [ ]:
ignore_index=True

In [ ]:
print(retail_df.shape)

(8494, 16)


In [ ]:
print(retail_df.shape)
retail_df.head()

(8494, 16)


,transaction_id,customer_id,customer_name,product_id,price,product_name,category,purchase_location,city,transaction_date,quantity,payment_method,discount,email,phone,payment_status
0,1,642,Troy Mitchell,108,15000.0,Mixer grinder,Home Appliances,offline,Bangalore,2025-12-06 00:00:00,3,UPI,0.10,Troy60@gmail.com,8385276968,successful
1,2,881,Sarah Guerrero,102,70000.0,Phone,ELEC,online,Bangalore,2025-07-21 00:00:00,5,Card,0.35,SarahGuerrero@yahoo.com,7147248911,successful
2,3,505,Samantha Hull,109,65000.0,Refrigerator,Home Appliances,offline,Chennai,2025-07-11 00:00:00,2,UPI,0.05,SamanthaHull@outlook.com,7415321565,successful
3,4,794,Gerald Cooper,106,80000.0,sofa,FURN,offline,Chennai,2025-10-06 00:00:00,2,UPI,0.05,GeraldCooper@gmail.com,9739354201,successful
4,5,864,Cameron Black,108,15000.0,Mixer grinder,home appliances,offline,Chennai,2025-12-12 00:00:00,1,Cash,0.05,CameronBlack@yahoo.com,9938169477,successful


In [ ]:
product_df.head()

,product_id,product_name,category,price
0,101,Laptop,Electronics,250000
1,102,Phone,Electronics,70000
2,103,Shirt,Clothing,1300
3,104,Shoes,Clothing,8000
4,105,TV,Electronics,45000


In [ ]:
product_df.columns

Index(['product_id', 'product_name', 'category', 'price'], dtype='object')

In [ ]:
print("Missing Prices Before:")
print(retail_df["price"].isnull().sum())

Missing Prices Before:
809


In [ ]:
price_lookup = product_df.set_index("product_id")["price"]

In [ ]:
retail_df["price"] = retail_df["price"].fillna(
    retail_df["product_id"].map(price_lookup)
)

In [ ]:
print("Missing Prices After:")
print(retail_df["price"].isnull().sum())

Missing Prices After:
0


Missing Price Treatment

Problem:
809 records contained missing prices.

Solution:
A product price lookup table was created using the Product Details reference dataset. Missing prices were imputed using the corresponding product_id.

Reason:
This ensured accurate revenue calculations and maintained consistency with the organization's master product data.

In [ ]:
print("Missing Prices Before:")
print(retail_df["price"].isnull().sum())

Missing Prices Before:
0


In [ ]:
price_lookup = product_df.set_index("product_id")["price"]

retail_df["price"] = retail_df["price"].fillna(
    retail_df["product_id"].map(price_lookup)
)

In [ ]:
print("Missing Prices After:")
print(retail_df["price"].isnull().sum())

Missing Prices After:
0


In [ ]:
print(retail_df["category"].unique())

['Home Appliances' 'ELEC' 'FURN' 'home appliances' 'Electronics'
 'electronics' 'CLOTH' 'clothing' 'Clothing' 'furniture' 'Furniture'
 'HOME']


In [ ]:
category_mapping = {
    "ELEC": "Electronics",
    "electronics": "Electronics",

    "CLOTH": "Clothing",
    "clothing": "Clothing",

    "FURN": "Furniture",
    "furniture": "Furniture",

    "home appliances": "Home Appliances",
    "HOME": "Home Appliances"
}

In [ ]:
retail_df["category"] = retail_df["category"].replace(category_mapping)

In [ ]:
print(retail_df["category"].unique())

['Home Appliances' 'Electronics' 'Furniture' 'Clothing']


Category Standardization

Problem:
The category column contained multiple representations of the same business category, such as ELEC, electronics, Electronics, FURN, furniture, and HOME.

Solution:
A category mapping dictionary was created and applied using Pandas replace() to standardize category values.

Reason:
This ensured accurate aggregation and reporting in Power BI by preventing duplicate business categories.

In [ ]:
print(retail_df["category"].unique())

['Home Appliances' 'Electronics' 'Furniture' 'Clothing']


In [ ]:
print(retail_df["product_name"].unique())

['Mixer grinder' 'Phone' 'Refrigerator' 'sofa' 'laptop' 'phone'
 'Microwave' 'PHONE' 'Shoes' 'SOFA' 'REFRIGERATOR' 'refrigerator' 'SHOES'
 'SHIRT' 'Tv' 'DINING TABLE' 'dining table' 'MICROWAVE' 'Shirt' 'shirt'
 'TV' 'shoes' 'Laptop' 'tv' 'MIXER GRINDER' 'microwave' 'mixer grinder'
 'Dining table' 'Sofa' 'LAPTOP']


In [ ]:
retail_df["product_name"] = retail_df["product_name"].str.strip().str.title()

In [ ]:
print(retail_df["product_name"].unique())

['Mixer Grinder' 'Phone' 'Refrigerator' 'Sofa' 'Laptop' 'Microwave'
 'Shoes' 'Shirt' 'Tv' 'Dining Table']


Product Name Standardization

Problem:
Product names were stored in different formats across source systems.

Solution:
Whitespace was removed using strip() and product names were converted to title case using title().

Reason:
This ensured consistent product representation and accurate product-level reporting.

In [ ]:
print(retail_df["transaction_date"].dtype)

object


In [ ]:
retail_df["transaction_date"] = pd.to_datetime(
    retail_df["transaction_date"],
    errors="coerce"
)

In [ ]:
print(retail_df["transaction_date"].dtype)

datetime64[ns]


In [ ]:
print(
    retail_df["transaction_date"].isnull().sum()
)

0


Date Standardization

Problem:
transaction_date was stored as object datatype.

Solution:
The column was converted to datetime using Pandas to_datetime() function.

Reason:
This enables time-based analysis, trend reporting, and Power BI date intelligence features.

In [ ]:
print(retail_df["transaction_date"].dtype)

retail_df["transaction_date"] = pd.to_datetime(
    retail_df["transaction_date"],
    errors="coerce"
)

print(retail_df["transaction_date"].dtype)

print(retail_df["transaction_date"].isnull().sum())

datetime64[ns]
datetime64[ns]
0


In [ ]:
retail_df[["email","phone"]].head()

,email,phone
0,Troy60@gmail.com,8385276968
1,SarahGuerrero@yahoo.com,7147248911
2,SamanthaHull@outlook.com,7415321565
3,GeraldCooper@gmail.com,9739354201
4,CameronBlack@yahoo.com,9938169477


In [ ]:
def mask_email(email):
    if pd.isnull(email):
        return email

    name, domain = email.split("@")
    return name[:2] + "***@" + domain

In [ ]:
retail_df["email"] = retail_df["email"].apply(mask_email)

In [ ]:
def mask_phone(phone):
    phone = str(phone)

    return phone[:2] + "******" + phone[-2:]

In [ ]:
retail_df["phone"] = retail_df["phone"].apply(mask_phone)

In [ ]:
retail_df[["email","phone"]].head()

,email,phone
0,Tr***@gmail.com,83******68
1,Sa***@yahoo.com,71******11
2,Sa***@outlook.com,74******65
3,Ge***@gmail.com,97******01
4,Ca***@yahoo.com,99******77


PII Masking

Problem:
Customer email addresses and phone numbers contain personally identifiable information (PII).

Solution:
Email addresses and phone numbers were masked using custom Python functions.

Reason:
This protects sensitive customer information and ensures compliance with data governance and privacy requirements.

In [ ]:
retail_df["revenue"] = (
    retail_df["price"]
    * retail_df["quantity"]
    * (1 - retail_df["discount"])
)

In [ ]:
retail_df[
    ["price", "quantity", "discount", "revenue"]
].head()

,price,quantity,discount,revenue
0,15000.0,3,0.10,40500.0
1,70000.0,5,0.35,227500.0
2,65000.0,2,0.05,123500.0
3,80000.0,2,0.05,152000.0
4,15000.0,1,0.05,14250.0


In [ ]:
total_revenue = retail_df["revenue"].sum()

print("Total Revenue:", total_revenue)

Total Revenue: 1233270150.0


In [ ]:
revenue_by_category = retail_df.groupby(
    "category"
)["revenue"].sum()

print(revenue_by_category)

category
Clothing            18021900.0
Electronics        717581250.0
Furniture          256294000.0
Home Appliances    241373000.0
Name: revenue, dtype: float64


In [ ]:
revenue_by_city = retail_df.groupby(
    "city"
)["revenue"].sum()

print(revenue_by_city)

city
Bangalore    226097660.0
Chennai      258323770.0
Delhi        255345590.0
Hyderabad    248035365.0
Mumbai       245467765.0
Name: revenue, dtype: float64


In [ ]:
total_orders = retail_df["transaction_id"].nunique()

print("Total Orders:", total_orders)

Total Orders: 8000


In [ ]:
total_customers = retail_df["customer_id"].nunique()

print("Total Customers:", total_customers)

Total Customers: 1963


In [ ]:
avg_order_value = total_revenue / total_orders

print("Average Order Value:", avg_order_value)

Average Order Value: 154158.76875


KPI Calculation

Revenue was calculated using:

Revenue = Price × Quantity × (1 - Discount)

The following KPIs were generated:

1. Total Revenue
2. Revenue by Category
3. Revenue by City
4. Total Orders
5. Total Customers
6. Average Order Value

These KPIs provide insights into business performance, customer activity, and regional sales trends.

In [ ]:
retail_df.to_csv(
    "cleaned_retail_data.csv",
    index=False
)

In [ ]:
retail_df["revenue"] = (
    retail_df["price"]
    * retail_df["quantity"]
    * (1 - retail_df["discount"])
)

total_revenue = retail_df["revenue"].sum()

print(total_revenue)

revenue_by_category = retail_df.groupby("category")["revenue"].sum()

print(revenue_by_category)

revenue_by_city = retail_df.groupby("city")["revenue"].sum()

print(revenue_by_city)

1233270150.0
category
Clothing            18021900.0
Electronics        717581250.0
Furniture          256294000.0
Home Appliances    241373000.0
Name: revenue, dtype: float64
city
Bangalore    226097660.0
Chennai      258323770.0
Delhi        255345590.0
Hyderabad    248035365.0
Mumbai       245467765.0
Name: revenue, dtype: float64


In [ ]:
total_orders = retail_df["transaction_id"].nunique()

total_customers = retail_df["customer_id"].nunique()

avg_order_value = total_revenue / total_orders

print("Total Orders:", total_orders)

print("Total Customers:", total_customers)

print("Average Order Value:", avg_order_value)

Total Orders: 8000
Total Customers: 1963
Average Order Value: 154158.76875


In [ ]:
top_products = retail_df.groupby(
    "product_name"
)["revenue"].sum().sort_values(
    ascending=False
)

print(top_products)

product_name
Laptop           497325000.0
Sofa             157744000.0
Phone            134232000.0
Refrigerator     123802250.0
Dining Table      98550000.0
Microwave         88911000.0
Tv                86024250.0
Mixer Grinder     28659750.0
Shoes             15615600.0
Shirt              2406300.0
Name: revenue, dtype: float64


In [ ]:
payment_analysis = retail_df.groupby(
    "payment_method"
)["revenue"].sum()

print(payment_analysis)

payment_method
Card          324104355.0
Cash          316414425.0
NetBanking    286542690.0
UPI           306208680.0
Name: revenue, dtype: float64


In [ ]:
retail_df["payment_status"].value_counts()

,count
payment_status,
successful,8000
failed,494


In [ ]:
retail_df.to_csv(
    "cleaned_retail_data.csv",
    index=False
)

print("Dataset Exported Successfully")

Dataset Exported Successfully


In [ ]:
import os

print(os.getcwd())
print(os.listdir())

/content
['.config', 'cleaned_retail_data.csv', 'USECASE - Data Engineering.xlsx', 'USECASE - Data Engineering (1).xlsx', 'sample_data']


In [ ]:
from google.colab import files

files.download("cleaned_retail_data.csv")

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>